datasetinsights/datasets/base.py

```diff
@@ -1,43 +1,38 @@
 from abc import ABCMeta, abstractmethod
 
 
-class Dataset(metaclass=ABCMeta):
-    """Abstract base class for datasets.
-    """
+class DatasetRegistry(ABCMeta):
+    registry = {}
+
+    def __new__(cls, name, bases, namespace):
+
+        if "name" not in namespace:
+            raise RuntimeError(
+                f"{name} must define a class-level dataset name!"
+            )
+        new_cls = super().__new__(cls, name, bases, namespace)
+        if namespace["name"] != "":
+            DatasetRegistry.registry[namespace["name"]] = new_cls
+        return new_cls
+
+    @classmethod
+    def find(cls, name):
+        dataset_cls = DatasetRegistry.registry.get(name)
+        if dataset_cls:
+            return dataset_cls
+        else:
+            raise ValueError(f"Dataset '{name}' does not exist:")
 
     @staticmethod
-    def create(name, **kwargs):
-        """Create a new instance of the dataset subclass
-
-        Args:
-            name (str): unique identifier for a dataset subclass
-            config (dict): parameters specific to each dataset subclass
-                used to create a dataset instance
+    def list_datasets():
+        return DatasetRegistry.registry.keys()
 
-        Returns:
-            an instance of the specified dataset subclass
-        """
-        dataset_cls = Dataset.find(name)
 
-        return dataset_cls(**kwargs)
+class Dataset(metaclass=DatasetRegistry):
+    """Abstract base class for datasets.
+    """
 
-    @staticmethod
-    def find(name):
-        """Find Dataset subclass based on the given name
-
-        Args:
-            name (str): unique identifier for a dataset subclass
-
-        Returns:
-            a label of the specified dataset subclass
-        """
-        dataset_classes = Dataset.__subclasses__()
-        dataset_names = [d.__name__ for d in dataset_classes]
-        if name in dataset_names:
-            dataset_cls = dataset_classes[dataset_names.index(name)]
-            return dataset_cls
-        else:
-            raise NotImplementedError(f"Unknown Dataset class {name}!")
+    name = ""
```

Comment:
Why do we need this?

Reply:
This is used to register datasets in the  DatasetRegistry metaclass. Since the class Dataset is a abstract base class we don't want to register it in the DatasetRegistry and have it be callable in the cli.

```
class DatasetRegistry(ABCMeta):
    registry = {}

    def __new__(cls, name, bases, namespace):

        if "name" not in namespace:
            raise RuntimeError(
                f"{name} must define a class-level dataset name!"
            )
        new_cls = super().__new__(cls, name, bases, namespace)
        if namespace["name"] != "":
            DatasetRegistry.registry[namespace["name"]] = new_cls
        return new_cls

```